jni_gen/test/jackson_core_test/third_party/src/dartjni.h

```diff
@@ -0,0 +1,172 @@
+#include <jni.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#if _WIN32
+#include <windows.h>
+#else
+#include <pthread.h>
+#include <unistd.h>
+#endif
+
+#if _WIN32
+#define FFI_PLUGIN_EXPORT __declspec(dllexport)
+#else
+#define FFI_PLUGIN_EXPORT
+#endif
+
+#if defined _WIN32
+#define thread_local __declspec(thread)
+#else
+#define thread_local __thread
+#endif
+
+#ifdef __ANDROID__
+#include <android/log.h>
+#endif
+
+#define JNI_LOG_TAG "Dart-JNI"
+
+#ifdef __ANDROID__
+#define __ENVP_CAST (JNIEnv **)
+#else
+#define __ENVP_CAST (void **)
+#endif
+
+struct jni_context {
+	JavaVM *jvm;
+	jobject classLoader;
+	jmethodID loadClassMethod;
+	jobject currentActivity;
+	jobject appContext;
+};
+
+extern thread_local JNIEnv *jniEnv;
+
+extern struct jni_context jni;
+
+enum DartJniLogLevel {
+	JNI_VERBOSE = 2,
+	JNI_DEBUG,
+	JNI_INFO,
+	JNI_WARN,
+	JNI_ERROR
+};
+
+FFI_PLUGIN_EXPORT struct jni_context GetJniContext();
+
+FFI_PLUGIN_EXPORT JavaVM *GetJavaVM(void);
+
+FFI_PLUGIN_EXPORT JNIEnv *GetJniEnv(void);
+
+FFI_PLUGIN_EXPORT JNIEnv *SpawnJvm(JavaVMInitArgs *args);
+
+FFI_PLUGIN_EXPORT jclass LoadClass(const char *name);
+
+FFI_PLUGIN_EXPORT jobject GetClassLoader(void);
+
+FFI_PLUGIN_EXPORT jobject GetApplicationContext(void);
+
+FFI_PLUGIN_EXPORT jobject GetCurrentActivity(void);
+
+FFI_PLUGIN_EXPORT void SetJNILogging(int level);
+
+FFI_PLUGIN_EXPORT jstring ToJavaString(char *str);
+
+FFI_PLUGIN_EXPORT const char *GetJavaStringChars(jstring jstr);
+
+FFI_PLUGIN_EXPORT void ReleaseJavaStringChars(jstring jstr, const char *buf);
+
+/// For use by jni_gen's generated code
```

Comment:
?

Reply:
Updated comment. These 2 are function pointer variables initialized by dart code to concrete implementations in `dartjni` base library.

This is because

* we don't know at compile time where in build artifacts folder `dartjni.so` exists -> No compile time linking.
* Default dlopen is RTLD_LOCAL on linux and android at least, thus the symbols are not automatically initialized from previously DLOpen'd library.

I am working around this by having dart code set whatever we need to lookup. I think using `dlsym` in C will become more complex and also probably less portable.

Reply:
👍 

I was just confused by the comment, thanks for clarifying.